### ISSUE TITLE
PID error

### ISSUE THREAD

**Issue by Kkaholoaa** [moiexpositoalonsolab/grenepipe#32]
Hi Lucas,
 
Tl;dr: What is a PID, and how can I amend my slurm profile to fix it?
--------------------------------------------------------------------------------------
To begin, Thank you for your “troubleshooting” page on the grenepipe wiki, as it helped me identify a specific error that was happening across all my samples.
 
The error is regarding PID, which (after a lot of google searches) means that the process ID is off, and it is occurring during my paired end read trimming step. Here is the error I got from the slurm log:
 
![image](https://github.com/moiexpositoalonsolab/grenepipe/assets/69033117/80717b02-e619-4d9b-953f-d9f014b1d311)

 
And here is the rule for trim_reads_pe for reference:
![image](https://github.com/moiexpositoalonsolab/grenepipe/assets/69033117/ece9c22c-a421-4750-883f-0fd8d6d77d1a)

 
*Problem Solving/Troubleshooting Attempt:*
 
After trying to understand the problem, I came to the conclusion that the problem may be with the slurm profile, as snakemake may have some sort of issue working with Stanford’s Server: Sherlock. Sherlock does in fact use slurm, and so I am convinced that I would need to amend grenepipe's provided slurm profile so that it can run properly. However, after doing some digging in grenepipe's slurm profile, including these these three files below, grenepipe does not have any instances where it describes a Process ID.
* slurm-status.py (This is the one that I think has the issue!!)
* slurm-jobscript.sh
* slurm-submit.py
 
Additionally, I do not think it is in slurm_utilis.py because it seems that file was meant not to be amended, so my gut tells me to leave that one alone.
 
Overall, my question for you is “What is a PID, and how can I amend my slurm profile to fix this bug?”
 
My best educated guess to resolve this is to input a code in one of those files that says
                if PID == “not found”:
                                PID = pid
 
What do you think?
Do you have any thought’s as if I’m heading in the right direction?
 
Thank you!
Kaiku Kaholoa'a

**Comment by lczech** [comment]
Hey Kaiku,

woooow you really dug deep there!

Your first question: The PID is the internal number that the operating system (at least Unix/Linux/Mac - bit different on Windows) uses to identify a process or program that you are running. It's just the way that programs are internally organized and addressed. If you want to end a process externally for example (because it might be stuck, or something), you can usually call `kill 1234` with the PID there.

Your second question: From what you wrote, I really have no idea where the problem is coming from. I don't think that filtering for existing PIDs in the slurm script is the way to go though. This seems to be an issue within the trimmomatic script itself. The cluster and slurm are not concerned with the internals of a rule once it is running. Or why did you come to the conclusion that slurm is involved at all?

Based on the error that you reported, I'd instead look into the trimmomatic rule, and see where and why it is breaking. The exact code for the rule execution is [here](https://snakemake-wrappers.readthedocs.io/en/0.74.0/wrappers/trimmomatic/pe.html), which you can copy into a script, and instead of having the rule execute the wrapper script, have it execute your copy of the script. You can then add debugging log messages to it to see how far it gets and where it breaks.

As a pre-test though to see if this is actually being caused by that particular rule, you could try to use another trimming tool, and see if that works. That would then hint at the trimmomatics script being the issue. If that however also fails with a similar error, then I'd say you are right in the assumption that the error is coming more from a snakemake/slurm thing.

Hope that helps, cheers
Lucas

**Comment by lczech** [comment]
Oh wait, there is more: 

 - If you have access to another cluster (Carnegie?), you could try that as well, and see if you get similar issues there.
 - Also, if you think slurm is the issue, you can instead get an interactive slurm session on the cluster with some resources, and run the pipeline within there - executing it with `--cores` instead of `--profile` so that it runs "locally" within that interactive session, instead of submitting the rules as jobs. Failure or success of that would also give strong hints into where this error is coming from.

Lucas

**Comment by Kkaholoaa** [comment]
Hi Lucas, 

Sorry for the late reply, I really wanted to try to troubleshoot this on my own but it seems I hit a wall - probably because of my lack of computational knowledge. The good news is I was able to identify that the issue seems to be with my wrappers, and I would really appreciate any advice you have for me to troubleshoot this...

To begin, I switched trimming tools from trimmomatic to adapterremoval, and it worked! At the same time, I also started receiving errors for fastqc: 

<img width="1000" alt="Screenshot 2023-05-19 at 1 10 24 AM" src="https://github.com/moiexpositoalonsolab/grenepipe/assets/69033117/48fa1e22-ae6b-410e-a33b-bdcd53227c43">


By looking into the fast qc rule, I found yet another PID error: 

<img width="1085" alt="Screenshot 2023-05-19 at 12 37 13 AM" src="https://github.com/moiexpositoalonsolab/grenepipe/assets/69033117/c03c2b64-2ad1-4132-b6fa-5447ee9a7e21">

Per your suggestion, I also ran grenepipe locally using --cores instead of --profile, and still came across the PID issues, meaning that it is probably not slurm that is causing this issue. 

As it turns out, both trimmomatic and fastqc use snakemake wrappers, while adapter removal does not (it uses a shell script instead). This tells me that the error might have something to do with snakemake wrappers, or the way that slurm is interacting with them. 

Here, the PID errors point to the snakemake wrapper in both Trimmomatic and Fastqc, but with adapter removal there is no snakemake wrapper and the trim was successful: 

FastQC:
<img width="827" alt="Screenshot 2023-05-19 at 1 11 49 AM" src="https://github.com/moiexpositoalonsolab/grenepipe/assets/69033117/5b396745-b659-4f01-ae15-b2dafd6b30e1">

Trimmomatic:
<img width="657" alt="Screenshot 2023-05-19 at 1 12 05 AM" src="https://github.com/moiexpositoalonsolab/grenepipe/assets/69033117/9dcceefb-3ddf-475f-8be5-8cb9237ac05f">

Adapter Removal:
<img width="864" alt="Screenshot 2023-05-19 at 1 12 17 AM" src="https://github.com/moiexpositoalonsolab/grenepipe/assets/69033117/e989752f-b7a1-469c-bead-03f93ebcef45">


To explore more into this issue, I attempted to use debugging codes by having print functions  (simply added in echo statements to get more info on why the wrappers dont work... but I don't really understand debugging just yet), so that hopefully I could identify the location in which in both wrappers (trimmomatic and fastqc) errors occurred. Although my debugging didnt work at all, I did find that grenepipe also began to map my trimmed reads from adapter removal onto my reference genome. This then led to the discovery that BWA also does not work, and with no surprise, yet another PID error that points to a snakemake wrapper:

<img width="1034" alt="Screenshot 2023-05-19 at 12 48 16 AM" src="https://github.com/moiexpositoalonsolab/grenepipe/assets/69033117/437d6a32-75f9-4cf0-a4cd-2e8ad77a5bc1">

As a result, I am *convinced* the error is with snakemake wrappers, but google does not do a very good job as to explaining the issue nor any potential ways to resolve it, and this is where I need your help. 

Here is some of my environment information, and I noticed that the version of my snakemake wrapper utils is not the version recommended on the snakemake website, and I also noticed that everything in grenepipe was created by conda forge except for snakemake, which was created by bio conda. 
<img width="530" alt="Screenshot 2023-05-19 at 1 02 58 AM" src="https://github.com/moiexpositoalonsolab/grenepipe/assets/69033117/7907b19b-12af-43a2-b479-e9246053e6ef">

I also found some links that may be helpful, where some people are saying that their cluster does not support internet access (which snakemake requires), and another different post that describes that subworkflows have been depreciated in favor of snakemake modules (unsure if this is relevant here).

https://stackoverflow.com/questions/76267968/snakemake-wrappers-not-working-on-slurm-cluster-compute-nodes-without-internet

https://snakemake.readthedocs.io/en/stable/snakefiles/modularization.html

Finally, I'd just thought to send this over just in case, but here is my snakemake code that I am using to run grenepipe... maybe there is an error on the way I am running grenepipe??? For context, I use tmux to create a new session, then I activate the grenepipe environment, and finally I use the follow code to run grenepipe from within the grenepipe directory, but pointing to my analysis directory where both my config.yaml and my sample_table.tsv files are located.

<img width="516" alt="Screenshot 2023-05-19 at 1 06 22 AM" src="https://github.com/moiexpositoalonsolab/grenepipe/assets/69033117/9938d02c-9426-4a1f-9588-3d13e6003907">

Overall, do you have any advice for me at this time? Anything would be appreciated into how to resolve this snakemake wrapper issue... 

Thank you!
Kaiku

**Comment by Kkaholoaa** [comment]
Hi Lucas, apologies for not making it clear, but the PID errors are from the slurm logs (see below), and the normal logs are coming up as empty.

Slurm log 19203864:
<img width="992" alt="Screenshot 2023-05-19 at 1 08 47 PM" src="https://github.com/moiexpositoalonsolab/grenepipe/assets/69033117/3fc07d1c-f3d9-42f8-99a2-2f19dc17e2cf">

Slurm log 19203847:
<img width="992" alt="Screenshot 2023-05-19 at 1 10 09 PM" src="https://github.com/moiexpositoalonsolab/grenepipe/assets/69033117/bd43d702-b9f3-4286-96c0-0677c24daf9e">

Logs > FastQC is filled with a bunch of empty logs:
<img width="998" alt="Screenshot 2023-05-19 at 1 11 04 PM" src="https://github.com/moiexpositoalonsolab/grenepipe/assets/69033117/506e3131-ca82-460a-a1ad-8af6a6c3e9fe">

I agree that it seems like something weird is happening on the back end, and hopefully there is somewhere else that I could look for hints? 

Please let me know what you think, 
Thank you!
Kaiku

**Comment by lczech** [comment]
Hm, so, this is not an issues with the wrappers (as per my answer above), and not caused by a tool in the scripts or wrappers failing directly with any usable error... It could be an issue with the cluster itself - some misconfiguration or limitation that we are not aware of. Could you try running it on the Carnegie cluster instead?

**Comment by Kkaholoaa** [comment]
Hello, after running this on the Carnegie cluster instead of Stanford's Sherlock Cluster, the pipeline now works!! 

![Screenshot 2023-05-23 at 1 49 17 PM](https://github.com/moiexpositoalonsolab/grenepipe/assets/69033117/70339c03-ae45-472b-897b-a1d29aa15dc2)

Although it works on the Carnegie Cluster, I will attempt to work with the Stanford IT staff to get this issue resolved on the Sherlock cluster, and will post our findings here just in case any other clusters out there may have the same issue...

Thank you so much! 
Best, 
Kaiku

**Comment by lczech** [comment]
Awesome, so then this does not seem to be a grenepipe issue per-se, but more related to some cluster infrastructure or configuration. I'll close the issue here then - but feel free to add further findings. As you said, it might be relevant for others as well. And we can re-open should it turn out to be related to grenepipe after all ;-)

Thanks again for your persistence, and curious to hear more on this story, if you ever find out what's going on there!

**Comment by lczech** [comment]
Awesome, so then this does not seem to be a grenepipe issue per-se, but more related to some cluster infrastructure or configuration. I'll close the issue here then - but feel free to add further findings. As you said, it might be relevant for others as well. And we can re-open should it turn out to be related to grenepipe after all ;-)

Thanks again for your persistence, and curious to hear more on this story, if you ever find out what's going on there!
